File: android/app/src/main/java/com/liive/ride/ui/RideActiveSheets.kt
package com.liive.ride.ui

import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.FastOutSlowInEasing
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import com.liive.ride.*
import com.liive.ride.designsystem.*

@Composable
fun RideMatchingSheet(state: RideUiState, onEvent: (RideEvent) -> Unit) {
    val c = LiiveTheme.colors
    val transition = rememberInfiniteTransition(label = "matching")

    LiiveBottomSheet(modifier = Modifier.testTag(RideTestTags.MatchingSheet)) {
        Column(
            Modifier
                .fillMaxWidth()
                .padding(
                    top = RideSheetLayout.matchingContentTopPadding,
                    bottom = RideSheetLayout.matchingContentBottomPadding
                ),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            Row(
                horizontalArrangement = Arrangement.spacedBy(RideSheetLayout.inlineGap),
                modifier = Modifier.padding(bottom = LiiveSpacing.l)
            ) {
                repeat(RideSheetLayout.matchingDotCount) { index ->
                    val dotProgress by transition.animateFloat(
                        initialValue = 0f,
                        targetValue = 1f,
                        animationSpec = infiniteRepeatable(
                            animation = tween(
                                durationMillis = RideSheetLayout.matchingDotDurationMs,
                                delayMillis = index * RideSheetLayout.matchingDotDelayMs,
                                easing = FastOutSlowInEasing
                            ),
                            repeatMode = RepeatMode.Reverse
                        ),
                        label = "matchingDot$index"
                    )
                    Box(
                        Modifier.offset(y = RideSheetLayout.matchingDotLift * dotProgress)
                            .alpha(0.5f + dotProgress * 0.5f)
                            .size(RideSheetLayout.matchingDotSize).clip(CircleShape).background(c.accent)
                    )
                }
            }
            Text("Finding your driver…", color = c.text, style = MaterialTheme.typography.headlineSmall)
            Text(
                "Matching you with a nearby${if (state.config.femaleOnly) " female-only" else ""} ${state.config.tier.name.lowercase()} driver and reserving a legal curb.",
                color = c.textSecondary,
                style = LiiveSheetMeta,
                textAlign = TextAlign.Center,
                modifier = Modifier
                    .widthIn(max = RideSheetLayout.matchingDescriptionMaxWidth)
                    .padding(top = RideSheetLayout.inlineGap)
            )
            Row(
                horizontalArrangement = Arrangement.spacedBy(RideSheetLayout.controlGap),
                modifier = Modifier.padding(top = LiiveSpacing.l)
            ) {
                LiiveBadge("Curb reserved", BadgeColor.Success, dot = true)
                if (state.config.femaleOnly) LiiveBadge("Female-only pool", BadgeColor.Accent)
            }
        }
        LiiveButton(
            title = "Cancel",
            onClick = { onEvent(RideEvent.CancelMatching) },
            fullWidth = true,
            variant = LiiveButtonVariant.Secondary,
            size = LiiveButtonSize.Lg,
            capsule = true
        )
    }
}

@Composable
fun RideEnrouteSheet(state: RideUiState, onEvent: (RideEvent) -> Unit) {
    val c = LiiveTheme.colors
    val config = state.config
    val driver = state.driver
    LiiveBottomSheet(modifier = Modifier.testTag(RideTestTags.EnrouteSheet)) {
        Row(Modifier.fillMaxWidth().padding(bottom = RideSheetLayout.headerBottomPadding)) {
            Text(
                state.tripSummary.enrouteTitle,
                color = c.text,
                style = MaterialTheme.typography.headlineSmall,
                modifier = Modifier.alignByBaseline()
            )
            Spacer(Modifier.weight(1f))
            Text(
                "to ${config.destinationName}",
                color = c.textSecondary,
                style = LiiveSheetMeta,
                modifier = Modifier.alignByBaseline()
            )
        }
        LiiveDriverCard(
            name = driver.name,
            rating = driver.rating,
            vehicle = driver.vehicle,
            plate = driver.plate,
            eta = state.tripSummary.driverEta,
            speaking = true
        ) {
            LiiveButton(
                title = "",
                onClick = {},
                variant = LiiveButtonVariant.Tinted,
                icon = RideIcons.Phone,
                iconOnly = true,
                contentDescription = "Call driver"
            )
        }
        state.tripSummary.transferStatus?.let { MultiLegPanel(it) }
        Row(
            Modifier.fillMaxWidth().padding(top = RideSheetLayout.sectionGap),
            horizontalArrangement = Arrangement.spacedBy(RideSheetLayout.rowGap)
        ) {
            LiiveButton(
                title = "Message",
                onClick = {},
                modifier = Modifier.weight(1f),
                variant = LiiveButtonVariant.Secondary,
                size = LiiveButtonSize.Lg,
                icon = RideIcons.Message
            )
            LiiveButton(
                title = "Cancel Ride",
                onClick = { onEvent(RideEvent.CancelRide) },
                modifier = Modifier.weight(1f),
                variant = LiiveButtonVariant.DestructivePlain,
                size = LiiveButtonSize.Lg
            )
        }
    }
}

@Composable
private fun MultiLegPanel(transferStatus: String) {
    val c = LiiveTheme.colors
    Column(
        Modifier
            .fillMaxWidth()
            .padding(top = RideSheetLayout.multiLegPanelTopPadding)
            .clip(LiiveRadius.lg)
            .background(c.surfaceRaised)
            .padding(RideSheetLayout.multiLegPanelPadding),
        verticalArrangement = Arrangement.spacedBy(RideSheetLayout.rowGap)
    ) {
        Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(RideSheetLayout.controlGap)) {
            Icon(painterResource(RideIcons.Map), null, tint = c.accent, modifier = Modifier.size(RideSheetLayout.multiLegIconSize))
            Text(
                "Multi-leg journey",
                color = c.text,
                style = MaterialTheme.typography.titleMedium,
                fontWeight = FontWeight.SemiBold
            )
        }
        LiiveProgressDots(legs = 2, current = 2)
        Box(Modifier.fillMaxWidth().height(RideSheetLayout.progressSeparatorTrackHeight)) {
            Box(
                Modifier.align(Alignment.BottomCenter)
                    .fillMaxWidth()
                    .height(RideSheetLayout.hairlineHeight)
                    .background(c.separator)
            )
        }
        Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(RideSheetLayout.inlineGap)) {
            Icon(painterResource(RideIcons.Walk), null, tint = c.warning, modifier = Modifier.size(RideSheetLayout.transferIconSize))
            Text(transferStatus, color = c.textSecondary, style = MaterialTheme.typography.bodySmall)
        }
    }
}

@Composable
fun RideCompleteSheet(state: RideUiState, onEvent: (RideEvent) -> Unit) {
    if (state.paid) PaidReceiptSheet(state, onEvent) else PaymentSheet(state, onEvent)
}

@Composable
fun RideSOSConfirmation(onEmergency: () -> Unit, onCancel: () -> Unit) {
    val c = LiiveTheme.colors
    Box(
        Modifier
            .fillMaxSize()
            .background(c.scrimStrong)
            .testTag(RideTestTags.SosConfirmation),
        contentAlignment = Alignment.Center
    ) {
        Column(
            Modifier
                .widthIn(max = RideSheetLayout.sosPanelMaxWidth)
                .clip(LiiveRadius.xl)
                .background(c.surface)
                .padding(RideSheetLayout.sosPanelPadding),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            Text("Emergency Alert", color = c.text, style = MaterialTheme.typography.headlineSmall)
            Text(
                "This will immediately alert emergency services and your emergency contacts. Are you sure?",
                color = c.textSecondary,
                style = MaterialTheme.typography.bodySmall,
                textAlign = TextAlign.Center,
                modifier = Modifier.padding(
                    top = RideSheetLayout.sosMessageTopPadding,
                    bottom = RideSheetLayout.sosMessageBottomPadding
                )
            )
            LiiveButton("Call Emergency Services", fullWidth = true, variant = LiiveButtonVariant.Destructive, size = LiiveButtonSize.Lg, capsule = true, onClick = onEmergency)
            Spacer(Modifier.height(RideSheetLayout.sosButtonGap))
            LiiveButton("Cancel", fullWidth = true, variant = LiiveButtonVariant.Plain, onClick = onCancel)
        }
    }
}

File: Packages/RideSharingFeature/Sources/RideSharingFeature/RideActiveSheets.swift
import SwiftUI

struct RideMatchingSheetView: View {
    @ObservedObject var viewModel: RideSharingViewModel
    @State private var animate = false

    private var config: RideConfiguration { viewModel.state.config }

    var body: some View {
        LiiveBottomSheet {
            VStack(spacing: 0) {
                HStack(spacing: 6) {
                    ForEach(0..<3, id: \.self) { index in
                        Circle()
                            .fill(LiiveColor.accent)
                            .frame(width: 9, height: 9)
                            .offset(y: animate ? -7 : 0)
                            .opacity(animate ? 1 : 0.5)
                            .animation(
                                .easeInOut(duration: 0.6)
                                    .repeatForever(autoreverses: true)
                                    .delay(Double(index) * 0.16),
                                value: animate
                            )
                    }
                }
                .padding(.bottom, 16)

                Text("Finding your driver…")
                    .liiveStyle(.title3)
                    .foregroundColor(LiiveColor.text)
                Text("Matching you with a nearby\(config.femaleOnly ? " female-only" : "") \(config.tier.rawValue) driver and reserving a legal curb.")
                    .font(Font.custom(LiiveFont.family, size: 14))
                    .foregroundColor(LiiveColor.textSecondary)
                    .multilineTextAlignment(.center)
                    .frame(maxWidth: 280)
                    .padding(.top, 6)

                HStack(spacing: 8) {
                    LiiveBadge("Curb reserved", color: .success, dot: true)
                    if config.femaleOnly {
                        LiiveBadge("Female-only pool", color: .accent)
                    }
                }
                .padding(.top, 16)
            }
            .frame(maxWidth: .infinity)
            .padding(.top, 8)
            .padding(.bottom, 22)

            LiiveButton("Cancel", variant: .secondary, size: .lg, shape: .capsule, fullWidth: true) {
                viewModel.handle(.cancelMatching)
            }
        }
        .accessibilityIdentifier(RideAccessibilityIdentifier.matchingSheet)
        .onAppear { animate = true }
    }
}

struct RideEnrouteSheetView: View {
    @ObservedObject var viewModel: RideSharingViewModel

    private var config: RideConfiguration { viewModel.state.config }
    private var driver: RideDriver { viewModel.state.driver }
    private var trip: RideTripSummary { viewModel.state.tripSummary }

    var body: some View {
        LiiveBottomSheet {
            HStack(alignment: .firstTextBaseline) {
                Text(trip.enrouteTitle)
                    .liiveStyle(.title3)
                    .foregroundColor(LiiveColor.text)
                Spacer()
                Text("to \(config.destinationName)")
                    .font(Font.custom(LiiveFont.family, size: 14))
                    .foregroundColor(LiiveColor.textSecondary)
            }
            .padding(.bottom, 12)

            LiiveDriverCard(
                name: driver.name,
                rating: driver.rating,
                vehicle: driver.vehicle,
                plate: driver.plate,
                eta: trip.driverETA,
                speaking: true
            ) {
                LiiveButton(
                    "",
                    variant: .tinted,
                    icon: Image(systemName: "phone.fill"),
                    iconOnly: true,
                    accessibilityLabel: "Call driver"
                ) {}
            }

            if let transferStatus = trip.transferStatus {
                multiLegPanel(transferStatus: transferStatus)
            }

            HStack(spacing: 10) {
                LiiveButton("Message", variant: .secondary, size: .lg, fullWidth: true, icon: Image(systemName: "message.fill")) {}
                LiiveButton("Cancel Ride", variant: .destructivePlain, size: .lg, fullWidth: true) {
                    viewModel.handle(.cancelRide)
                }
            }
            .padding(.top, 14)
        }
        .accessibilityIdentifier(RideAccessibilityIdentifier.enrouteSheet)
    }

    private func multiLegPanel(transferStatus: String) -> some View {
        VStack(alignment: .leading, spacing: 10) {
            HStack(spacing: 8) {
                Image(systemName: "map.fill")
                    .font(.system(size: 16, weight: .semibold))
                    .foregroundColor(LiiveColor.accent)
                Text("Multi-leg journey")
                    .font(LiiveFont.subhead.weight(.semibold))
                    .foregroundColor(LiiveColor.text)
            }
            LiiveProgressDots(legs: 2, current: 2)
            Rectangle()
                .fill(LiiveColor.separator)
                .frame(height: 0.5)
                .padding(.top, 2)
            HStack(spacing: 6) {
                Image(systemName: "figure.walk")
                    .font(.system(size: 15, weight: .semibold))
                    .foregroundColor(LiiveColor.warning)
                Text(transferStatus)
                    .font(LiiveFont.footnote)
                    .foregroundColor(LiiveColor.textSecondary)
            }
        }
        .padding(14)
        .background(LiiveColor.surfaceRaised)
        .clipShape(RoundedRectangle(cornerRadius: LiiveRadius.lg, style: .continuous))
        .padding(.top, 12)
    }
}

struct RideCompleteSheetView: View {
    @ObservedObject var viewModel: RideSharingViewModel

    private var config: RideConfiguration { viewModel.state.config }
    private var driver: RideDriver { viewModel.state.driver }
    private var fare: RideFareBreakdown { config.fareBreakdown }
    private var trip: RideTripSummary { viewModel.state.tripSummary }

    var body: some View {
        if viewModel.state.paid {
            paidReceipt
        } else {
            paymentSheet
        }
    }

    private var paidReceipt: some View {
        LiiveBottomSheet {
            VStack(spacing: 0) {
                LiiveIconCircle(systemName: "checkmark", color: .success, size: 56, filled: true)
                Text("Thanks for riding")
                    .liiveStyle(.title2)
                    .foregroundColor(LiiveColor.text)
                    .padding(.top, 14)
                Text("\(fare.total.ridePrice) paid to \(driver.firstName) · receipt sent")
                    .font(LiiveFont.subhead.monospacedDigit())
                    .foregroundColor(LiiveColor.textSecondary)
                    .padding(.top, 6)
            }
            .frame(maxWidth: .infinity)
            .padding(.vertical, 10)

            LiiveButton("Done", variant: .primary, size: .lg, shape: .capsule, fullWidth: true) {
                viewModel.handle(.reset)
            }
            .padding(.top, 20)
        }
        .accessibilityIdentifier(RideAccessibilityIdentifier.receiptSheet)
    }

    private var paymentSheet: some View {
        LiiveBottomSheet {
            HStack(spacing: 10) {
                LiiveIconCircle(systemName: "flag.fill", color: .success, size: 36, filled: true)
                VStack(alignment: .leading, spacing: 2) {
                    Text("You've arrived")
                        .liiveStyle(.title3)
                        .foregroundColor(LiiveColor.text)
                    Text("\(config.destinationName) · \(trip.completedDuration) · \(trip.completedDistance)")
                        .font(LiiveFont.footnote.monospacedDigit())
                        .foregroundColor(LiiveColor.textSecondary)
                }
                Spacer()
            }
            .padding(.bottom, 14)

            VStack(spacing: 0) {
                LiiveFareRow(label: "Ride fare", amount: fare.rideFare.ridePrice)
                LiiveFareRow(label: "Tax & fees", amount: fare.taxAndFees.ridePrice)
                if let credit = fare.costShareCredit {
                    LiiveFareRow(label: "Cost-share credit", amount: "–\(credit.ridePrice)", muted: true)
                }
                Rectangle()
                    .fill(LiiveColor.separator)
                    .frame(height: 0.5)
                LiiveFareRow(label: "Total", amount: fare.total.ridePrice, total: true)
            }
            .padding(.horizontal, 14)
            .padding(.top, 8)
            .padding(.bottom, 14)
            .background(LiiveColor.surfaceRaised)
            .clipShape(RoundedRectangle(cornerRadius: LiiveRadius.lg, style: .continuous))
            .padding(.bottom, 12)

            VStack(spacing: 0) {
                LiiveListRow(title: "Apple Pay", value: "default", divider: false, chevron: true) {
                    LiiveIconCircle(systemName: "applelogo", color: .neutral, size: 32)
                }
            }
            .background(LiiveColor.surfaceRaised)
            .clipShape(RoundedRectangle(cornerRadius: LiiveRadius.lg, style: .continuous))
            .padding(.bottom, 12)

            ratingControl
            LiiveButton("Pay \(fare.total.ridePrice)", variant: .primary, size: .lg, shape: .capsule, fullWidth: true, tabularNumbers: true) {
                viewModel.handle(.pay)
            }
            Text("Secured by Stripe")
                .font(LiiveFont.caption1)
                .foregroundColor(LiiveColor.textTertiary)
                .frame(maxWidth: .infinity)
                .padding(.top, 10)
        }
        .accessibilityIdentifier(RideAccessibilityIdentifier.paymentSheet)
    }

    private var ratingControl: some View {
        VStack(spacing: 8) {
            Text("Rate your driver")
                .font(Font.custom(LiiveFont.family, size: 14))
                .foregroundColor(LiiveColor.textSecondary)
            HStack(spacing: 6) {
                ForEach(1...5, id: \.self) { value in
                    Button(action: { viewModel.handle(.rate(value)) }) {
                        Image(systemName: "star.fill")
                            .font(.system(size: 28))
                            .foregroundColor(value <= viewModel.state.rating ? LiiveColor.star : LiiveColor.fill)
                            .padding(2)
                    }
                    .buttonStyle(.plain)
                }
            }
        }
        .padding(.bottom, 16)
    }
}

private extension RideDriver {
    var firstName: String {
        name.split(separator: " ").first.map(String.init) ?? name
    }
}
